From d09ac8c3ef4159434e15a4793474be49ac64d082 Mon Sep 17 00:00:00 2001
From: Dongwon Kim <dongwon.kim@intel.com>
Date: Thu, 24 Jul 2025 11:56:02 -0700
Subject: [PATCH] hw/virtio-gpu: Cursor size can be changed

HW cursor image size is set to 128, 128 in windows guest (using ZC
driver). QEMU should be able to resize (via reallocation) the container
of guest mouse cursor to accomodate this new size.

Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
---
 hw/display/virtio-gpu.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/hw/display/virtio-gpu.c b/hw/display/virtio-gpu.c
index cf84c54ac3..418c0598dd 100644
--- a/hw/display/virtio-gpu.c
+++ b/hw/display/virtio-gpu.c
@@ -15,6 +15,7 @@
 #include "qemu/units.h"
 #include "qemu/iov.h"
 #include "sysemu/cpus.h"
+#include <math.h>
 #include "ui/console.h"
 #include "ui/rect.h"
 #include "trace.h"
@@ -57,9 +58,14 @@ void virtio_gpu_update_cursor_data(VirtIOGPU *g,
     }
 
     if (res->blob_size) {
-        if (res->blob_size < (s->current_cursor->width *
-                              s->current_cursor->height * 4)) {
-            return;
+        if (res->blob_size != (s->current_cursor->width *
+                               s->current_cursor->height * 4)) {
+            /* resize 'current_cursor' for the new blob_size
+             * The assumption is the cursor image is square.
+             */
+            int width = sqrt(res->blob_size / 4);
+            g_free(s->current_cursor);
+            s->current_cursor = cursor_alloc(width, width);
         }
         data = res->blob;
     } else if (res->image) {
-- 
2.43.0

